Always use legacy layouts

slavapestov opened the pull request:
The layouts of resilient value types shipped in the Swift 5 standard library and overlays will forever be frozen in time for backward deployment to old Objective-C runtimes. This PR ensures that even if the layouts of these types evolve in the future, binaries built to run on the old runtime will continue to lay out class instances in a manner compatible with Swift 5.

Fixes <rdar://problem/45646886>.

slavapestov committed:
Copy legacy layouts to stdlib build directory

This adds a new copy-legacy-layouts-${platform}-${arch} target for each
platform and architecture that the standard library is built for.

If the platform and architecture has a corresponding layout file in
stdlib/public/legacy_layouts/${platform}/layouts-${arch}.yaml, the
target copies this file to the build directory; otherwise, it does
nothing.

When building Swift code, the subroutines in SwiftSource.cmake add a
dependency on this target from each Swift code target.

Finally, we ensure that the YAML files are copied into the toolchain
package when building a toolchain.

slavapestov commented:
@jrose-apple This isn't quite ready (has some test failures, don't ask) but I would appreciate a quick review.

slavapestov commented:
Ok, I also need to make sure we include the YAML files when building the toolchain package.

jrose-apple commented on lib/IRGen/GenType.cpp:
```diff
@@ -1125,13 +1126,17 @@ TypeConverter::TypeConverter(IRGenModule &IGM)
 
+  auto libraryImportPath = IGM.Context.SearchPathOpts.RuntimeLibraryImportPath;
+  llvm::SmallString<128> defaultPath(libraryImportPath);
+  llvm::sys::path::append(defaultPath, "layouts.yaml");
```
You can't assume this'll always be there; it won't on Linux, for example.

slavapestov replied:
Should we enumerate somewhere which platforms have this and which ones do not? I don't want to just silently continue if the file is not present, in case it gets lost in the installation process somehow.

jrose-apple replied:
Sure, that sounds fine to me. (Naming comment still applies.)

slavapestov committed:
IRGen: Rename -read-type-info-path flag to -read-legacy-type-info-path

slavapestov committed:
Use -read-legacy-type-info-path in Runtime/class-layout-from-objc.m test

slavapestov committed:
IRGen: Always use YAML files for completely fragile class layout

The layouts of resilient value types shipped in the Swift 5 standard library
x and overlays will forever be frozen in time for backward deployment to old
Objective-C runtimes. This PR ensures that even if the layouts of these types
evolve in the future, binaries built to run on the old runtime will continue
to lay out class instances in a manner compatible with Swift 5.

Fixes <rdar://problem/45646886>.

slavapestov committed:
IRGen: Stop using 'completely fragile' mode when emitting class metadata

This was a stop-gap measure until the YAML legacy type layout infrastructure
was in place.

slavapestov committed:
IRGen: Remove -enable-class-resilience staging flag

jrose-apple commented on include/swift/AST/IRGenOptions.h:
```diff
@@ -168,18 +168,20 @@ class IRGenOptions {
+  /// Normally if the -read-legacy-type-info flag is not specified, we look for
+  /// a file named "legacy.yaml" in the same directory as Swift.swiftmodule.
+  /// Passing this flag completely disables this behavior.
+  unsigned DisableLegacyTypeInfo : 1;
```
Comment needs updating.

jrose-apple commented on include/swift/AST/IRGenOptions.h:
```diff
@@ -168,18 +168,20 @@ class IRGenOptions {
+
   /// The path to load legacy type layouts from.
-  StringRef ReadTypeInfoPath;
+  StringRef ReadLegacyTypeInfo;
```
This is still a path, no? Right now this doesn't sound like the name of a string. (Neither does the new command-line option name.)

slavapestov replied:
I'm going to rename it to `-read-legacy-type-info-path`, but I'm keeping `-disable-legacy-type-info`.

jrose-apple commented on lib/IRGen/GenType.cpp:
```diff
@@ -1114,6 +1116,30 @@ TypeConverter::getLegacyTypeInfo(NominalTypeDecl *decl) const {
   return found->second;
 }
 
+static const char *platformsWithLegacyLayouts[][2] = {
```
Nitpick: this can use `llvm::StringLiteral`, and then it'll be faster to compare.

slavapestov replied:
It's only compared once per IRGenModule creation so I don't think it's performance-sensitive, but I'll see if it cleans it up anyway.

jrose-apple commented on lib/IRGen/GenType.cpp:
```diff
@@ -1125,13 +1151,32 @@ TypeConverter::TypeConverter(IRGenModule &IGM)
+  // suite. To prevent these from failing when the user hasn't build the
+  // standard library for that target, don't even try loading the legacy
+  // type info when -parse-stdlib is specified.
+  if (IGM.IRGen.Opts.DisableLegacyTypeInfo)
```
This comment doesn't match the test.

jrose-apple commented on lib/IRGen/GenType.cpp:
```diff
@@ -1125,13 +1151,32 @@ TypeConverter::TypeConverter(IRGenModule &IGM)
+
+  StringRef path = IGM.IRGen.Opts.ReadLegacyTypeInfo;
+  if (path.empty())
+    path = defaultPath;
```
Nitpick: Seems like the whole `defaultPath` computation could go in this `if`.

jrose-apple commented on stdlib/CMakeLists.txt:
```diff
@@ -107,6 +107,18 @@ swift_create_stdlib_targets("swift-stdlib" "sibopt" TRUE)
+        "${SWIFT_SOURCE_DIR}/stdlib/public/legacy_layouts/${platform}/layouts-${arch}.yaml"
+        "${SWIFTLIB_DIR}/${platform}/layouts-${arch}.yaml")
+  endforeach()
+endforeach()
```
You need to filter out platforms and architectures that don't have legacy layouts here.

slavapestov replied:
Do you want me to duplicate the list or are you OK with a check that the source file exists?

jrose-apple replied:
I think the source file existing is okay for the CMake side. The compiler will complain if one's missing.

jrose-apple reviewed (commented):
General approach for turning this on seems fine. I still have a bunch of little comments.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

gottesmm commented on test/IRGen/completely_fragile_class_layout.sil:
```diff
@@ -1,11 +1,16 @@
 // RUN: %target-swift-frontend  -I %t -emit-ir -enable-resilience %s -read-legacy-type-info-path=%S/Inputs/legacy_type_info/a.yaml | %FileCheck %s --check-prefix=CHECK --check-prefix=CHECK-%target-ptrsize -DINT=i%target-ptrsize
 
 // RUN: %target-swift-frontend  -I %t -emit-ir -enable-resilience -O %s
+=======
```
(Just a drive by... but it looks like you messed up a merge here.

jrose-apple commented on stdlib/CMakeLists.txt:
```diff
@@ -107,6 +107,31 @@ swift_create_stdlib_targets("swift-stdlib" "sibopt" TRUE)
+  foreach(arch ${SWIFT_SDK_${sdk}_ARCHITECTURES})
+    set(platform "${SWIFT_SDK_${sdk}_LIB_SUBDIR}")
+    set(input "${SWIFT_SOURCE_DIR}/stdlib/public/legacy_layouts/${platform}/layouts-${arch}.yaml")
+    set(output "${SWIFTLIB_DIR}/${platform}/layouts-${arch}.yaml")
```
This'll copy them into the build directory, but I think you still need a `swift_install_in_components` for it to show up in a toolchain. (Test with "swift-ci build toolchain".)

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

jrose-apple reviewed (approved):
Assuming the built toolchain works this looks good to me.

slavapestov commented:
The source compatibility suite uses the built toolchain.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
(Why only a smoke test? Source compat passed and so did full test on OS X, since then I've only made a test change for Linux)

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test
